crates/astria-sequencer/src/address/mod.rs

```diff
@@ -0,0 +1,103 @@
+use anyhow::ensure;
+use astria_core::primitive::v1::{
+    Address,
+    AddressError,
+    ADDRESS_LEN,
+};
+#[cfg(not(test))]
+pub(crate) use regular::*;
+#[cfg(test)]
+pub(crate) use testonly::*;
+
+pub(crate) fn base_prefixed(arr: [u8; ADDRESS_LEN]) -> Address {
+    Address::builder()
+        .array(arr)
+        .prefix(get_base_prefix())
+        .try_build()
+        .expect("the prefix must have been set as a valid bech32 prefix, so this should never fail")
+}
+
+pub(crate) fn try_base_prefixed(slice: &[u8]) -> Result<Address, AddressError> {
+    Address::builder()
+        .slice(slice)
+        .prefix(get_base_prefix())
+        .try_build()
+}
+
+pub(crate) fn ensure_base_prefix(address: &Address) -> anyhow::Result<()> {
+    ensure!(
+        get_base_prefix() == address.prefix(),
+        "address has prefix `{}` but only `{}` is permitted",
+        address.prefix(),
+        crate::address::get_base_prefix(),
+    );
+    Ok(())
+}
+
+#[cfg(not(test))]
+mod regular {
+    //! Logic to be used for a normal debug or release build of sequencer.
+
+    use std::sync::OnceLock;
+
+    use anyhow::Context as _;
+
+    static BASE_PREFIX: OnceLock<String> = OnceLock::new();
+
+    pub(crate) fn initialize_base_prefix(base_prefix: &str) -> anyhow::Result<()> {
+        // construct a dummy address to see if we can construct it; fail otherwise.
+        try_construct_dummy_address_from_prefix(base_prefix)
+            .context("failed constructing a dummy address from the provided prefix")?;
+
+        BASE_PREFIX.set(base_prefix.to_string()).expect(
+            "THIS IS A BUG: the base prefix was already more than once; it must only be \
+             initialized once and upon receiving an init-chain consensus request. It can not be \
+             initialized twice or concurrently from more than one task or thread.",
+        );
+
+        Ok(())
+    }
+
+    pub(crate) fn get_base_prefix() -> &'static str {
+        BASE_PREFIX
+            .get()
+            .expect(
+                "the base prefix must have been set during chain init; if not set, the chain was \
```

Comment:
```suggestion
                "the base prefix must have been set during `InitChain`; if not set, the chain was \
```

This is the official ABCI method unlikely to change, might be easier to trace when called out.